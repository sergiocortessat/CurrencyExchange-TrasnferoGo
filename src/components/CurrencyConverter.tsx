import React, { useEffect } from "react";
import { useSelector, useDispatch } from "react-redux";
import {
  setFromCurrency,
  setToCurrency,
  setAmount,
  setConvertedAmount,
  setDisableButton,
} from "../redux/actions";
import { AppState } from "../redux/reducers";
import Button from "./UI/Button";
import FormControl from "./UI/FormControl";
import Select from "./UI//Select";
import TextField from "./UI//TextInput";
import DisplayConversionRate from "./DisplayConversionRate";
import InputLabel from "./UI//Input";
import TextBlock from "./UI//TextBlock";
import fetchConversion from "./FetchConversion";
import { SupportedCurrencies } from "../data/currencies";
import { SwapHoriz } from "@material-ui/icons";
import { IconButton } from "@material-ui/core";
import "../Custom.scss";

const CurrencyConverter: React.FC = () => {
  const {
    fromCurrency,
    toCurrency,
    amount,
    convertedAmount,
    conversionRate,
    disableButton,
  } = useSelector((state: AppState) => state);

  const dispatch = useDispatch();

  useEffect(() => {
    fetchConversion(amount, fromCurrency, toCurrency, dispatch);
  }, [amount, fromCurrency, toCurrency, dispatch]);

  const handleConvert = () => {
    fetchConversion(amount, fromCurrency, toCurrency, dispatch);
    dispatch(setDisableButton(true));
  };

  const handleAmountChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const newAmount = e.target.value;
    dispatch(setAmount(newAmount));
    if (newAmount === "") {
      dispatch(setConvertedAmount(""));
    } else if (conversionRate !== 0) {
      const converted = parseFloat(newAmount) * conversionRate;
      dispatch(setConvertedAmount(converted.toString()));
    }
  };

  const handleConvertedToChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const newConvertedAmount = e.target.value;
    dispatch(setConvertedAmount(newConvertedAmount));
    if (newConvertedAmount === "") {
      dispatch(setAmount(""));
    } else if (conversionRate !== 0) {
      const amount = parseFloat(newConvertedAmount) / conversionRate;
      dispatch(setAmount(amount.toString()));
    }
  };

  const handleFromCurrencyChange = (
    e: React.ChangeEvent<{ value: unknown }>
  ) => {
    dispatch(setFromCurrency(e.target.value as string));
  };

  const handleToCurrencyChange = (e: React.ChangeEvent<{ value: unknown }>) => {
    dispatch(setToCurrency(e.target.value as string));
  };

  const handleExchangeClick = () => {
    dispatch(setFromCurrency(toCurrency));
    dispatch(setToCurrency(fromCurrency));
    if (disableButton) {
      dispatch(setAmount(convertedAmount));
      dispatch(setConvertedAmount(amount));
    }
  };

  return (
    <div className="section">
      <div className="page">
        <div className="form-control">
          <FormControl className="form-control--from-currency-label">
            <InputLabel htmlFor="from-currency" className="from-currency-label">From</InputLabel>
            <Select
              id="from-currency"
              value={fromCurrency}
              onChange={handleFromCurrencyChange}
              values={SupportedCurrencies}
              className="from-currency-select"
            />
          </FormControl>
          <div className="exchange-button-container">
            <IconButton
              className="exchange-button"
              onClick={handleExchangeClick}
            >
              <SwapHoriz />
            </IconButton>
          </div>
          <FormControl className="form-control--to-currency-label">
            <InputLabel htmlFor="to-currency" className="to-currency-label">To</InputLabel>
            <Select
              id="to-currency"
              value={toCurrency}
              onChange={handleToCurrencyChange}
              values={SupportedCurrencies}
              className="to-currency-select"
            />
          </FormControl>
        </div>
        <div className="text-input">
          <TextField
            id="amount"
            label="Amount"
            value={amount}
            onChange={handleAmountChange}
            className="text-input-amount"
            min={0}
            max={1000}
            adornments={true}
            currency={fromCurrency}
          />
          {disableButton && (
            <TextField
              id="converted-to"
              label="Converted to"
              value={convertedAmount}
              onChange={handleConvertedToChange}
              className="text-input-converted-amount"
              min={0}
              max={1000}
              adornments={true}
              currency={toCurrency}
            />
          )}
        </div>
        {disableButton && (
          <DisplayConversionRate
            conversionRate={conversionRate}
            className="text-field--rate"
            currencyFrom={fromCurrency}
            currencyTo={toCurrency}
          />
        )}
        {disableButton && (
          <TextBlock
            className="text-block"
            type="normal"
          >
            <span>All figures are live mid-market rates, wich are for information purposes only.</span>
            <br />
            <span>To see the rates for money transfer, please select money send options</span>
          </TextBlock>
        )}
        {!disableButton && (
          <div className="button-container">
            <Button
              variant="contained"
              color="primary"
              onClick={handleConvert}
              disabled={parseInt(amount) >= 1}
              className="full-width-button button"
            >
              Convert
            </Button>
          </div>
        )}
      </div>
    </div>
  );
};

export default CurrencyConverter;

