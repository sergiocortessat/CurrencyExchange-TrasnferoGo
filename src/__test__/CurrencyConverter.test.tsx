import React from "react";
import { render, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import CurrencyConverter from "../components/CurrencyConverter";
import store from "../redux/store";

describe("CurrencyConverter", () => {
  it("renders correctly", () => {
    const {asFragment} = render(
      <Provider store={store}>
        <CurrencyConverter />
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
    expect(screen.getByText("Convert")).toBeInTheDocument();
    expect(screen.getByText("From")).toBeInTheDocument();
    expect(screen.getByText("To")).toBeInTheDocument();
  });
});
