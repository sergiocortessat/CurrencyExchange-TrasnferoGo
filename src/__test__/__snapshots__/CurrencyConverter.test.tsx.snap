// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrencyConverter renders correctly 1`] = `
<DocumentFragment>
  <div
    class="section"
  >
    <div
      class="page"
    >
      <div
        class="form-control"
      >
        <div
          class="MuiFormControl-root form-control--from-currency-label"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root from-currency-label MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
            data-shrink="true"
            for="from-currency"
          >
            From
          </label>
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline from-currency-select MuiInputBase-formControl MuiInput-formControl"
          >
            <div
              aria-haspopup="listbox"
              aria-labelledby="from-currency"
              class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
              id="from-currency"
              role="button"
              tabindex="0"
            >
              <img
                alt="EUR"
                src="EUR.png"
              />
              EUR
            </div>
            <input
              aria-hidden="true"
              class="MuiSelect-nativeInput"
              tabindex="-1"
              value="EUR"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSelect-icon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
          </div>
        </div>
        <div
          class="exchange-button-container"
        >
          <button
            class="MuiButtonBase-root MuiIconButton-root exchange-button"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M6.99 11L3 15l3.99 4v-3H14v-2H6.99v-3zM21 9l-3.99-4v3H10v2h7.01v3L21 9z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="MuiFormControl-root form-control--to-currency-label"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root to-currency-label MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
            data-shrink="true"
            for="to-currency"
          >
            To
          </label>
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline to-currency-select MuiInputBase-formControl MuiInput-formControl"
          >
            <div
              aria-haspopup="listbox"
              aria-labelledby="to-currency"
              class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
              id="to-currency"
              role="button"
              tabindex="0"
            >
              <img
                alt="GBP"
                src="GBP.png"
              />
              GBP
            </div>
            <input
              aria-hidden="true"
              class="MuiSelect-nativeInput"
              tabindex="-1"
              value="GBP"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSelect-icon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
          </div>
        </div>
      </div>
      <div
        class="text-input"
      >
        <div
          class="MuiFormControl-root MuiTextField-root text-input-amount"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
            data-shrink="true"
            for="amount"
            id="amount-label"
          >
            Amount
          </label>
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedEnd"
              id="amount"
              type="number"
              value="1.00"
            />
            <div
              class="MuiInputAdornment-root MuiInputAdornment-positionEnd"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
              >
                EUR
              </p>
            </div>
          </div>
        </div>
      </div>
      <div
        class="button-container"
      >
        <button
          class="custom-button contained primary-color full-width-button button custom-button--false"
        >
          Convert
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
